ktor-io/common/src/io/ktor/utils/io/ByteReadChannelOperations.kt

```diff
@@ -568,53 +563,92 @@ public suspend fun ByteReadChannel.readUntil(
     check(matchString.size > 0) {
         "Empty match string not permitted for readUntil"
     }
+
     val partialMatchTable = buildPartialMatchTable(matchString)
-    var matchIndex = 0
-    val matchBuffer = ByteArray(matchString.size)
+    val partialMatchBuffer = Buffer()
+    val input = readBuffer
     var rc = 0L
+    var matchIndex = 0
 
-    suspend fun appendPartialMatch() {
-        writeChannel.writeFully(matchBuffer, 0, matchIndex)
-        rc += matchIndex
-        matchIndex = 0
+    fun checkBounds(extra: Long) {
+        if (rc + extra > limit) {
+            throw IOException(
+                "Limit of $limit bytes exceeded " +
+                    "while searching for \"${matchString.toSingleLineString()}\""
+            )
+        }
     }
 
-    fun resetPartialMatch(byte: Byte) {
-        while (matchIndex > 0 && byte != matchString[matchIndex]) {
-            matchIndex = partialMatchTable[matchIndex - 1]
+    // Quick scan for first byte
+    suspend fun advanceToNextMatch() {
+        while (!input.exhausted() || awaitContent()) {
+            val nextMatch = input.indexOf(matchString[0])
+            when {
+                nextMatch == -1L -> {
+                    checkBounds((input as Buffer).size)
+                    rc += input.transferTo(writeChannel.writeBuffer)
+                }
+                else -> {
+                    checkBounds(nextMatch)
+                    rc += input.readAtMostTo(writeChannel.writeBuffer as Buffer, nextMatch)
+                    return
+                }
+            }
         }
     }
 
-    while (!isClosedForRead) {
-        val byte = readByte()
+    // Slow sequential check for potential match
+    suspend fun readCompleteMatch(): Boolean {
+        while (!input.exhausted() || awaitContent()) {
+            val byte = input.readByte()
 
-        if (matchIndex > 0 && byte != matchString[matchIndex]) {
-            appendPartialMatch()
-            resetPartialMatch(byte)
-        }
+            if (matchIndex > 0 && byte != matchString[matchIndex]) {
+                // update match index from our table
+                val oldMatchIndex = matchIndex
+                while (matchIndex > 0 && byte != matchString[matchIndex]) {
+                    matchIndex = partialMatchTable[matchIndex - 1]
```

Comment:
Initially, the `matchIndex` is 0, and that's the only place I found where it's reassigned. So, how is that variable updated?